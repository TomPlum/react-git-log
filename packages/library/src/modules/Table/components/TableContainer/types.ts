import { CSSProperties } from 'react'
import { CustomTableRow } from 'modules/Table'

export interface TableContainerProps {
  rowQuantity: number
  className?: string
  row?: CustomTableRow
  styleOverrides?: CSSProperties
}