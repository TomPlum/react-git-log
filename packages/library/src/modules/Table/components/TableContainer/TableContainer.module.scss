.tableContainer {
  display: grid;
  grid-template-columns: minmax(350px, 4fr) minmax(100px, 1fr) 195px;
}