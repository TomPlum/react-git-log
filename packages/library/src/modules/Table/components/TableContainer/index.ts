export * from './types'
export * from './TableContainer'