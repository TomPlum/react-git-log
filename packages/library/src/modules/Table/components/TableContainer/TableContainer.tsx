import classNames from 'classnames'
import styles from './TableContainer.module.scss'
import { PropsWithChildren, useMemo } from 'react'
import { useGitContext } from 'context/GitContext'
import { placeholderCommits } from 'modules/Graph/hooks/usePlaceholderData/data'
import { ROW_HEIGHT } from 'constants/constants'
import { HEADER_ROW_HEIGHT, TABLE_MARGIN_TOP } from 'modules/Table/constants'
import { TableContainerProps } from 'modules/Table/components/TableContainer/types'

export const TableContainer = ({
  rowQuantity,
  children,
  className,
  styleOverrides
}: PropsWithChildren<TableContainerProps>) => {
  const { rowSpacing, showHeaders } = useGitContext()

  const gridTemplateRows = useMemo(() => {
    // If no commits are visible as we're showing
    // the placeholder data for the skeleton view,
    // then use that size, else just use the log data length.
    const commitsVisible = rowQuantity > 0
      ? rowQuantity
      : placeholderCommits.length

    // If the table headers are turned off, then we simply
    // repeat the same row height for all rows.
    if (!showHeaders) {
      return `repeat(${commitsVisible}, ${ROW_HEIGHT}px)`
    }

    // With no row spacing, the header row height lines
    // up with the first data row fine. But when the row
    // spacing is increased, we must subtract half of it
    // from the height of the first header row to counteract
    // the gap between the header and the first data row.
    const headerRowHeight = HEADER_ROW_HEIGHT - (rowSpacing / 2)

    // All other rows (with data) get a fixed height.
    const remainingRowsHeight = `repeat(${commitsVisible}, ${ROW_HEIGHT}px)`

    return `${headerRowHeight}px ${remainingRowsHeight}`
  }, [rowQuantity, rowSpacing, showHeaders])

  const marginTop = useMemo(() => {
    if (showHeaders) {
      return TABLE_MARGIN_TOP
    }

    return (rowSpacing / 2) + TABLE_MARGIN_TOP
  }, [rowSpacing, showHeaders])
  
  return (
    <div
      id='react-git-log-table'
      data-testid='react-git-log-table'
      style={{
        ...styleOverrides,
        marginTop,
        gridTemplateRows,
        rowGap: rowSpacing
      }}
      className={classNames(styles.tableContainer, className)}
    >
      {children}
    </div>
  )
}