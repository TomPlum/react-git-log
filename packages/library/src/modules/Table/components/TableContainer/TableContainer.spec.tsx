import { render } from '@testing-library/react'
import { TableContainer } from 'modules/Table/components/TableContainer/TableContainer'

describe('Table Container', () => {
  it('should render a wrapping div element with the correct styles when a custom row is used', () => {
    const customRow = () => (
      <div>
        <p>Custom Row Contents</p>
      </div>
    )

    const { asFragment } = render(
      <TableContainer rowQuantity={5} row={customRow}>
        <div>
          Stubbed Table Contents
        </div>
      </TableContainer>
    )

    expect(asFragment()).toMatchSnapshot()
  })
})