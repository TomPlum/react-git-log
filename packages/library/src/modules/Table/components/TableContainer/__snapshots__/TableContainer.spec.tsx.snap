// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table Container > should render a wrapping div element with the correct styles when a custom row is used 1`] = `
<DocumentFragment>
  <div
    data-testid="react-git-log-table"
    id="react-git-log-table"
    style="margin-top: 12px;"
  >
    <div>
      Stubbed Table Contents
    </div>
  </div>
</DocumentFragment>
`;
